src/WinRT.Runtime/Marshalers.cs

```diff
@@ -18,21 +20,92 @@ public static void Dispose(this GCHandle handle)
                 handle.Free();
             }
         }
+
+#if !NET
+        public static unsafe ref readonly char GetPinnableReference(this string str)
+        {
+            fixed (char* p = str)
+            {
+                return ref *p;
+            }
+        }
+#endif
     }
 
-    // TODO: minimize heap allocations for marshalers by eliminating explicit try/finally
-    // and adopting ref structs with non-IDisposable Dispose and 'using var ...' pattern,
-    // as well as passing marshalers to FromAbi by ref so they can be conditionally disposed.
 #if EMBED
     internal
 #else 
     public
 #endif
     class MarshalString
     {
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct HSTRING_HEADER
+        {
+            IntPtr reserved1;
+            int reserved2;
+            int reserved3;
+            int reserved4;
+            int reserved5;
+        }
+
         private IntPtr _header;
-        public GCHandle _gchandle;
-        public IntPtr _handle;
+        private GCHandle _gchandle;
+
+        public ref struct Pinnable
+        {
+            private HSTRING_HEADER _header;
+            private string _value;
+#if DEBUG
+            private bool _pinned;
+#endif
+
+            public Pinnable(string value)
+            {
+                _value = value ?? "";
+                _header = default;
+#if DEBUG
+                _pinned = false;
+#endif            
+            }
+
+            public ref readonly char GetPinnableReference()
+            {
+#if DEBUG
+                _pinned = true;
+#endif
+                return ref _value.GetPinnableReference();
+            }
+
+            public unsafe IntPtr GetAbi()
+            {
+#if DEBUG
+                // We assume that the string is pinned by the calling code
+                Debug.Assert(_pinned);
+#endif
+                if (_value == "") 
+                {
+                    return IntPtr.Zero;
+                }
+                IntPtr hstring;
+                Marshal.ThrowExceptionForHR(Platform.WindowsCreateStringReference(
+                    (char*)Unsafe.AsPointer(ref Unsafe.AsRef(in GetPinnableReference())),
+                    _value.Length,
+                    (IntPtr*)Unsafe.AsPointer(ref _header),
+                    &hstring));
+                return hstring;
+            }
+        }
+
+        public static Pinnable CreatePinnable(string value) => new(value);
+
+        public static IntPtr GetAbi(ref Pinnable p) => p.GetAbi();
+
+        public MarshalString(string value)
+        {
+            _gchandle = GCHandle.Alloc(value, GCHandleType.Pinned);
+            _header = Marshal.AllocHGlobal(Unsafe.SizeOf<HSTRING_HEADER>());

```

Comment:
Nit: If this `Marshal.AllocHGlobal` fails, you will end up with leaking GCHandle. There is not going to be anything to free the GCHandle. 

Just pointing it out - feel free to won't fix. Memory leaks like this are not a problem in practice. If the process fails to allocate 20 bytes of memory, it is guaranteed to crash very soon.

Reply:
no reason to alloc in the ctor - if I delay it to GetAbi, then the ctor can only fail if the GCHandle isn't allocated.